core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -657,6 +679,14 @@ class BlockManagerMasterEndpoint(
     }
   }
 
+  private def getMergerLocations(
+      numMergersNeeded: Int,
+      hostsToFilter: Set[String]): Seq[BlockManagerId] = {
+    // Copying the merger locations to a list so that the original mergerLocations won't be shuffled
+    val mergers = mergerLocations.values.filterNot(x => hostsToFilter.contains(x.host)).toSeq
+    Utils.randomize(mergers).take(numMergersNeeded)
```

Comment:
this is randomizing the locations, does it make sense to favor the current active executors? I guess perhaps the yarn scheduler backend could do extra checking if it so decided?  Perhaps this api should be more generic about getting block managers vs merger locations

Reply:
Are you preferring active executors to bias towards the possibility of increasing the chance of one or more active executors on those nodes when reducers start @tgravescs ? Prioritizing active executors does increase the chances of that ... did we do experiments with that sort of prioritization Venkat ?
I agree with the other points @venkata91 raised - just that with what Tom mentioned, there is a higher probability of having existing executors there already, and so not necessarily need to wait for allocation to kick in.

Reply:
Hm. We haven't really done much of experiments at this point. But this is an interesting area to explore further. Another thing we can possibly do is pass the merger locations information to `ExecutorAllocationManager` as part of `ShuffleMapStage` creation to give some sort of preference to these executors when we remove executors between the stages.

Reply:
For now, I have avoided shuffling the merger locations in `BlockManagerMasterEndpoint` but on a second thought we are only sending the `mergersNeeded` which can be less than the overall `shuffleMergers` available. Shouldn't we shuffle in `BlockManagerMasterEndpoint` itself as it has the whole list and what we are returning is the only required number of hosts. We shuffled the hosts info also to prevent any unnecessary node hotspots. @tgravescs Thoughts?

Reply:
yeah the main thing I was thinking was you already have these executors so you don't have to worry about requesting them from YARN and them potentially being full or not available at that point.    If you have the same number of reducers as mappers then you don't even need to ask YARN for different executors as you potentially already have them.  Now you do obviously have the potential that you send the data to an executor and then it idle times out anyway because another map is the long tail and takes a long time.
I get that the shuffle is trying to get away from hot spots and I obviously have no data on this. It would definitely be interesting to look at more though. Perhaps shuffle active ones, grab the number you need, if that isn't enough grab from  nonactive. That makes your tracking more complex though.  

I think in your experiments you were getting pretty good locality in production right?  Its going to depend on how busy your clusters are though too. Perhaps this is just a file a separate issue to investigate.

Reply:
After having offline discussions with @mridulm decided to make the change to prefer active executors. In the worst case, we would still get the behavior of without any preference.